instrumentation/log/http_output.go

```diff
@@ -0,0 +1,95 @@
+package log
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/pkg/errors"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+)
+
+type row struct {
+	level     string
+	timestamp time.Time
+	message   string
+	fields    []*Field
+}
+
+type httpOutput struct {
+	formatter LogFormatter
+	writer    io.Writer
+
+	bulkSize int
+	delay    time.Duration
+
+	lock *sync.Mutex
+	logs []*row
+}
+
+func (out *httpOutput) Append(level string, message string, fields ...*Field) {
+	row := &row{level, time.Now(), message, fields}
+
+	out.lock.Lock()
+	out.logs = append(out.logs, row)
+	out.lock.Unlock()
+
+	out.flush()
+}
+
+func (out *httpOutput) flush() {
+	out.lock.Lock()
+	defer out.lock.Unlock()
+
+	if len(out.logs) >= out.bulkSize {
+		b := new(bytes.Buffer)
+
+		for _, row := range out.logs {
+			b.Write([]byte(out.formatter.FormatRow(row.timestamp, row.level, row.message, row.fields...)))
+			b.Write([]byte("\n"))
+		}
+
+		out.logs = nil
+
+		go func() {
+			if n, err := out.writer.Write(b.Bytes()); err != nil {
+				fmt.Println(fmt.Sprintf("%s failed to send logs via http, %d bytes lost: %s", time.Now().String(), n, err))
+			}
+		}()
+	}
+}
+
+func NewHttpOutput(writer io.Writer, formatter LogFormatter, bulkSize int) Output {
+	return &httpOutput{
+		formatter: formatter,
+		writer:    writer,
+		bulkSize:  bulkSize,
+		lock:      &sync.Mutex{},
+	}
+}
+
+type httpWriter struct {
+	url string
+}
+
+func (w *httpWriter) Write(p []byte) (n int, err error) {
+	reader := bytes.NewReader(p)
```

Comment:
now, let's find a way to write to http via streaming as opposed to allocating a big byte array

Reply:
It's already memory-efficient since we allocate a buffer once and then just write to it. There is no double allocation here.

Reply:
what am I missing? I see that you create a new(bytes.Buffer) each time you flush the logs

Reply:
Yes, it happens on every flush, but before that there was double allocation with `strings.Join`, and now it's gone.